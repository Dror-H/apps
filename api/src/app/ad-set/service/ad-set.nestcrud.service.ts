import { Inject, Injectable } from '@nestjs/common';
import { TypeOrmCrudService } from '@nestjsx/crud-typeorm';
import { AdSet } from '../data/ad-set.entity';
import { AdSetRepository } from '../data/ad-set.repository';

@Injectable()
export class AdSetNestCrudService extends TypeOrmCrudService<AdSet> {
  constructor(@Inject(AdSetRepository) repo: AdSetRepository) {
    super(repo);
  }
}
