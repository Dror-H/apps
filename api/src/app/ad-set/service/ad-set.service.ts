import {
  AdSetCreationDTO,
  WorkspaceDTO,
  AdSetDTO,
  ResponseSuccess,
  SupportedProviders,
  TargetingDto,
  CacheTtlSeconds,
} from '@instigo-app/data-transfer-object';
import { ThirdPartyAdSetApiService } from '@instigo-app/third-party-connector';
import { CACHE_MANAGER, Inject, Injectable } from '@nestjs/common';
import { Cache } from 'cache-manager';
import { groupBy } from 'lodash';
import { from, Observable } from 'rxjs';
import { map, mergeMap, switchMap } from 'rxjs/operators';
import { AdSet } from '../data/ad-set.entity';
import { AdSetRepository } from '../data/ad-set.repository';

@Injectable()
export class AdSetService {
  @Inject(ThirdPartyAdSetApiService)
  private readonly thirdPartyAdSetApiService: ThirdPartyAdSetApiService;

  @Inject(AdSetRepository)
  private readonly repo: AdSetRepository;

  @Inject(CACHE_MANAGER)
  private readonly cacheManager: Cache;

  async create(adSet: Partial<AdSetCreationDTO>, currentWorkspace: WorkspaceDTO): Promise<AdSetDTO> {
    const { provider, adAccount, campaign } = adSet;
    const { accessToken } = currentWorkspace.owner.getAccessToken({ provider });
    const builtAdSet = await this.thirdPartyAdSetApiService.create({
      accessToken,
      adAccountProviderId: adAccount.providerId,
      provider,
      adSet: adSet,
    });
    builtAdSet.adAccount = adAccount;
    builtAdSet.campaign = campaign;
    return this.repo.save(builtAdSet);
  }

  bulkPatch(options: { adSets: Partial<AdSetDTO>[]; workspace: WorkspaceDTO }): Observable<any> {
    const { adSets, workspace } = options;
    void this.cacheManager.reset();
    return from(Object.entries(groupBy(adSets, 'provider'))).pipe(
      mergeMap(([provider, entries]) => {
        const { accessToken } = workspace.owner.getAccessToken({ provider });
        return this.thirdPartyAdSetApiService.bulkPatch({ provider, accessToken, adSets: entries });
      }),
      switchMap(() => this.repo.save(adSets)),
      map(() => new ResponseSuccess('AdSets successfully updated')),
    );
  }

  async findAdSetTargeting(options: {
    workspace: WorkspaceDTO;
    provider: SupportedProviders;
    adSetId: string;
  }): Promise<TargetingDto> {
    const { adSetId, provider, workspace } = options;
    const cacheKey = `adSetTargeting-${provider}-${adSetId}`;
    let result = await this.cacheManager.get(cacheKey);
    if (result) {
      return result as any;
    }
    const { accessToken } = workspace.owner.getAccessToken({ provider });
    result = await this.thirdPartyAdSetApiService.findAdSetTargeting({
      accessToken,
      adSetId,
      provider,
    });
    await this.cacheManager.set(cacheKey, result, { ttl: CacheTtlSeconds.ONE_HOUR });
    return result as any;
  }
}
