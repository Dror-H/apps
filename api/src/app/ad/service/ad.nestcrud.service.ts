import { Injectable } from '@nestjs/common';
import { TypeOrmCrudService } from '@nestjsx/crud-typeorm';
import { Ad } from '../data/ad.entity';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';

@Injectable()
export class AdNestCrudService extends TypeOrmCrudService<Ad> {
  constructor(
    @InjectRepository(Ad)
    public repo: Repository<Ad>,
  ) {
    super(repo);
  }
}
