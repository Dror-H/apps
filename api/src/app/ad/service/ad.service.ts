import { CACHE_MANAGER, Inject, Injectable } from '@nestjs/common';
import { AdCreationDTO, AdDTO, AdUpdateDTO, ResponseSuccess, WorkspaceDTO } from '@instigo-app/data-transfer-object';
import { Ad } from '@api/ad/data/ad.entity';
import { ThirdPartyAdApiService } from '@instigo-app/third-party-connector';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { Cache } from 'cache-manager';
import { from, Observable } from 'rxjs';
import { groupBy } from 'lodash';
import { map, mergeMap, switchMap } from 'rxjs/operators';

@Injectable()
export class AdService {
  @Inject(ThirdPartyAdApiService)
  private readonly thirdPartyAdApiService: ThirdPartyAdApiService;

  @InjectRepository(Ad)
  private readonly adRepository: Repository<Ad>;

  @Inject(CACHE_MANAGER)
  private readonly cacheManager: Cache;

  async create(ad: Partial<AdCreationDTO<any>>, currentWorkspace: WorkspaceDTO): Promise<AdDTO> {
    const { provider, adAccount } = ad;
    const { accessToken } = currentWorkspace.owner.getAccessToken({ provider });
    const builtAd = await this.thirdPartyAdApiService.create({
      accessToken,
      adAccountProviderId: adAccount.providerId,
      provider,
      ad,
    });
    builtAd.adAccount = adAccount;
    return this.adRepository.save(builtAd);
  }

  async bulkCreate(ads: Partial<AdCreationDTO<any>>[], currentWorkspace: WorkspaceDTO): Promise<Ad[]> {
    const { provider, adAccount, campaign } = ads[0];
    const { accessToken } = currentWorkspace.owner.getAccessToken({ provider });
    const createdAds = await this.thirdPartyAdApiService.createMany({
      accessToken,
      adAccountProviderId: adAccount.providerId,
      provider,
      ads: ads,
    });
    const adSet = ads[0].providerSpecificFields?.adSet;
    const mappedAds = createdAds.map((item) => ({ ...item, adAccount, campaign, adSet }));
    return this.adRepository.save(mappedAds);
  }

  public async updateOne(ad: AdUpdateDTO, currentWorkspace: WorkspaceDTO): Promise<Ad> {
    await this.cacheManager.reset();
    const { accessToken } = currentWorkspace.owner.getAccessToken({ provider: ad.provider });
    await this.thirdPartyAdApiService.updateAd({ provider: ad.provider, accessToken, ad });
    const newAd = {
      id: ad.id,
      name: ad.name,
      objectStorySpec: { ...ad.data, adTemplateType: ad.adTemplateType },
    } as Ad;
    return await this.adRepository.save(newAd);
  }

  bulkPatch(options: { ads: Partial<AdDTO>[]; workspace: WorkspaceDTO }): Observable<any> {
    const { ads, workspace } = options;
    void this.cacheManager.reset();
    return from(Object.entries(groupBy(ads, 'provider'))).pipe(
      mergeMap(([provider, entries]) => {
        const { accessToken } = workspace.owner.getAccessToken({ provider });
        return this.thirdPartyAdApiService.bulkPatch({ provider, accessToken, ads: entries });
      }),
      switchMap(() => this.adRepository.save(ads)),
      map(() => new ResponseSuccess('Ads successfully updated')),
    );
  }
}
