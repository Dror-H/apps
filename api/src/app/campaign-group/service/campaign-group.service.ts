import { Inject, Injectable } from '@nestjs/common';
import { TypeOrmCrudService } from '@nestjsx/crud-typeorm';
import { CampaignGroup } from '../data/campaign-group.entity';
import { CampaignGroupRepository } from '../data/campaign-group.repository';

@Injectable()
export class CampaignGroupNestCrudService extends TypeOrmCrudService<CampaignGroup> {
  constructor(@Inject(CampaignGroupRepository) public repo: CampaignGroupRepository) {
    super(repo);
  }
}
