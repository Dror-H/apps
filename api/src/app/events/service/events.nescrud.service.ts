import { Inject, Injectable } from '@nestjs/common';
import { TypeOrmCrudService } from '@nestjsx/crud-typeorm';
import { EventsRepository } from '@api/events/data/events.repository';
import { EventsEntity } from '@api/events/data/events.entity';

@Injectable()
export class EventsNestCrudService extends TypeOrmCrudService<EventsEntity> {
  constructor(@Inject(EventsRepository) eventsRepository: EventsRepository) {
    super(eventsRepository);
  }
}
