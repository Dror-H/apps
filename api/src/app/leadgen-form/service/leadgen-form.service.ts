import { Inject, Injectable } from '@nestjs/common';
import { TypeOrmCrudService } from '@nestjsx/crud-typeorm';
import { LeadgenForm } from '../data/leadgen-form.entity';
import { LeadgenFormRepository } from '../data/leadgen_form.repository';

@Injectable()
export class LeadgenFormNestCrudService extends TypeOrmCrudService<LeadgenForm> {
  constructor(@Inject(LeadgenFormRepository) public repo: LeadgenFormRepository) {
    super(repo);
  }
}
