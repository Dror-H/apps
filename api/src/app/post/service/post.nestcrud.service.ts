import { Inject, Injectable } from '@nestjs/common';
import { TypeOrmCrudService } from '@nestjsx/crud-typeorm';
import { Post } from '../data/post.entity';
import { PostRepository } from '../data/post.repository';

@Injectable()
export class PostNestCrudService extends TypeOrmCrudService<Post> {
  constructor(@Inject(PostRepository) repo: PostRepository) {
    super(repo);
  }
}
