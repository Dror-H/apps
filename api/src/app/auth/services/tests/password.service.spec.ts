import { MockRepository } from '@instigo-app/api-shared';
import { MailerService } from '@nestjs-modules/mailer';
import { ConfigService } from '@nestjs/config';
import { Test } from '@nestjs/testing';
import * as faker from 'faker';
import { PasswordService } from '../password.service';

describe('PasswordService', () => {
  let service: PasswordService;
  let mailerService: MailerService;
  let config: ConfigService;
  let userRepository: any;
  let authService: any;
  let prettyLinkService: any;

  beforeEach(async () => {
    const module = await Test.createTestingModule({
      providers: [
        PasswordService,
        { provide: 'ConfigService', useValue: {} },
        { provide: 'MailerService', useValue: {} },
        { provide: 'AuthService', useValue: {} },
        { provide: 'UserRepository', useValue: new MockRepository() },
        { provide: 'PrettyLinkService', useValue: {} },
      ],
    }).compile();

    service = module.get<PasswordService>(PasswordService);
    mailerService = module.get('MailerService');
    config = module.get('ConfigService');
    userRepository = module.get('UserRepository');
    authService = module.get('AuthService');
    prettyLinkService = module.get('PrettyLinkService');
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  it(`sendResetPasswordMail - should throw error when the mail couldn't be sent`, async () => {
    const email = faker.datatype.uuid();
    const password = faker.datatype.uuid();
    const user = { email, password };
    const options = { user: { email } };
    mailerService.sendMail = jest.fn().mockRejectedValue({ rejected: ['rejectedMail'] });
    userRepository.findByEmail = jest.fn().mockResolvedValue(user);
    config.get = jest.fn().mockReturnValue('support@email.com');
    prettyLinkService.passwordResetShortUrl = jest.fn().mockReturnValue({});
    // Assert
    await expect(service.sendResetPasswordMail(options)).rejects.toThrow(
      `Could not send reset password email: ${email}`,
    );
  });

  it(`sendResetPasswordMail - should throw error when user could't be found by emai`, async () => {
    const error = 'User Not found';
    userRepository.findByEmail = jest.fn().mockRejectedValue(() => {
      throw new Error(error);
    });
    // Assert
    await expect(service.sendResetPasswordMail({ user: { email: 'example@email.com' } })).rejects.toThrow(error);
  });

  it('sendResetPasswordMail - should send email', async () => {
    // Arrange
    const email = faker.datatype.uuid();
    const password = faker.datatype.uuid();
    const user = { email, password };
    const expectedResult = `Reset password email successfully sent: ${user.email}`;
    userRepository.findByEmail = jest.fn().mockResolvedValue(user);
    service.buildMailerOptions = jest.fn().mockReturnValue({});
    mailerService.sendMail = jest.fn().mockResolvedValue({});
    // Act
    const result = await service.sendResetPasswordMail({ user });
    // Assert
    expect(mailerService.sendMail).toHaveBeenCalled();
    expect(result).toEqual({ message: expectedResult });
  });

  it('resetPassword - it should reset password', async () => {
    // Arrange
    const email = faker.datatype.uuid();
    const options = { email, password: faker.datatype.uuid() };
    const expectedResult = { email, password: faker.datatype.uuid() };
    userRepository.findByEmail = jest.fn().mockResolvedValue(options);
    userRepository.save = jest.fn().mockResolvedValue(expectedResult);
    // Act
    const user = await service.resetPassword(options);
    // Assert
    expect(user).toEqual(expectedResult);
  });

  it('resetPassword - it should fail, email not found', async () => {
    // Arrange
    const email = faker.datatype.uuid();
    const options = { email, password: faker.datatype.uuid() };
    const error = 'User Not found';
    userRepository.findByEmail = jest.fn().mockRejectedValue(() => {
      throw new Error(error);
    });
    // Act and Assert
    await expect(service.resetPassword(options)).rejects.toThrow(error);
  });

  it('buildMailerOptions - should create mailer options', async () => {
    // Arrange
    const email = 'lupu60@gmail.com';
    const token = 'dasndsadknsan32321321';
    const username = 'lupu60';
    const from = 'moish@insticore.com';
    const BACKEND_HOST = 'http://instigo.api.io';
    const shortUrl = 'https://shorturl.at/bhiY4';

    config.get = jest.fn().mockReturnValueOnce(from).mockReturnValueOnce(BACKEND_HOST);
    prettyLinkService.passwordResetShortUrl = jest.fn().mockReturnValueOnce(shortUrl);
    const result = await service.buildMailerOptions({ userEmail: email, token, name: username });
    // Assert
    expect(result).toMatchSnapshot();
  });

  it('verifyResetPasswordToken - should return new token', async () => {
    // Arrange
    const expectedResult = faker.datatype.uuid();
    const token = `eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VyIjp7ImlkIjoiNTQzZjcxOTctM2VhMS00YmM5LWEwNjQtYTdjYThmMGYyYmY1IiwidXNlcm5hbWUiOm51bGwsImZpcnN0TmFtZSI6bnVsbCwibGFzdE5hbWUiOm51bGwsImVtYWlsIjoiZHJhZ29pdXJhdWwyMDA5QGdtYWlsLmNvbSIsImlzQWN0aXZlIjp0cnVlLCJwYXNzd29yZCI6IiQyYSQxMCQ4cGtuTmlpRlQ1TE0uVXlNTEJPeGN1MjdRQk5QWVp1TW81SUZjZWZ3S09Od0NiM3ZYQUZ2aSIsImNyZWF0ZWRBdCI6IjIwMjAtMDMtMThUMTQ6NTg6MDAuODk1WiIsInVwZGF0ZWRBdCI6IjIwMjAtMDMtMThUMTg6MzA6MzUuODIzWiIsImVtYWlsVmVyaWZpY2F0aW9uIjpmYWxzZSwib25ib2FyZGluZyI6e30sInJvbGVzIjpbIlVTRVIiXX0sInNjb3BlIjoicmVzZXRQYXNzd29yZEVtYWlsIiwiaWF0IjoxNTg0NjIwNDA4fQ.BcuXi7Q6QYJKjPMeTVotE97ym1K5t4Auz2WN532ygsw`;
    const options = { token };
    const email = 'dragoiuraul2009@gmail.com';
    const password = '$2a$10$8pknNiiFT5LM.UyMLBOxcu27QBNPYZuMo5IFcefwKONwCb3vXAFvi';
    const user = { email, password };
    userRepository.findByEmail = jest.fn().mockResolvedValue(user);
    authService.createJwtToken = jest.fn().mockResolvedValue({ token: expectedResult });
    // Act
    const result = await service.verifyResetPasswordToken(options);
    // Assert
    expect(result).toEqual({ token: expectedResult });
  });

  it('verifyResetPasswordToken - should throw JsonWebTokenError when password is changed ', async () => {
    // Arrange
    const expectedResult = faker.datatype.uuid();
    const token = `eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VyIjp7ImlkIjoiNTQzZjcxOTctM2VhMS00YmM5LWEwNjQtYTdjYThmMGYyYmY1IiwidXNlcm5hbWUiOm51bGwsImZpcnN0TmFtZSI6bnVsbCwibGFzdE5hbWUiOm51bGwsImVtYWlsIjoiZHJhZ29pdXJhdWwyMDA5QGdtYWlsLmNvbSIsImlzQWN0aXZlIjp0cnVlLCJwYXNzd29yZCI6IiQyYSQxMCQ4cGtuTmlpRlQ1TE0uVXlNTEJPeGN1MjdRQk5QWVp1TW81SUZjZWZ3S09Od0NiM3ZYQUZ2aSIsImNyZWF0ZWRBdCI6IjIwMjAtMDMtMThUMTQ6NTg6MDAuODk1WiIsInVwZGF0ZWRBdCI6IjIwMjAtMDMtMThUMTg6MzA6MzUuODIzWiIsImVtYWlsVmVyaWZpY2F0aW9uIjpmYWxzZSwib25ib2FyZGluZyI6e30sInJvbGVzIjpbIlVTRVIiXX0sInNjb3BlIjoicmVzZXRQYXNzd29yZEVtYWlsIiwiaWF0IjoxNTg0NjIwNDA4fQ.BcuXi7Q6QYJKjPMeTVotE97ym1K5t4Auz2WN532ygsw`;
    const options = { token };
    const email = 'dragoiuraul2009@gmail.com';
    const password = 'changed password';
    const user = { email, password };
    userRepository.findByEmail = jest.fn().mockResolvedValue(user);
    authService.createJwtToken = jest.fn().mockResolvedValue({ token: expectedResult });
    // Act and Assert
    await expect(service.verifyResetPasswordToken(options)).rejects.toThrow('invalid signature');
  });
});
