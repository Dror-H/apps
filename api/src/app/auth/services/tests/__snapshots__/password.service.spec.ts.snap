// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordService buildMailerOptions - should create mailer options 1`] = `
Object {
  "context": Object {
    "action_url": "https://shorturl.at/bhiY4",
    "name": "lupu60",
  },
  "subject": "Reset password email - Instigo",
  "template": "reset-password-email",
  "to": "lupu60@gmail.com",
}
`;
